src/plan_agg_bookend.c

```diff
@@ -0,0 +1,715 @@
+	{
+		Aggref	   *aggref = (Aggref *) node;
+
+		/* See if the Aggref should be replaced by a Param */
+		if (context->mm_path != NULL &&
+			list_length(aggref->args) == 2)
+		{
+			TargetEntry *curTarget = (TargetEntry *) linitial(aggref->args);
+			ListCell   *cell;
+
+			foreach(cell, context->mm_path->mmaggregates)
+			{
+				MinMaxAggInfo *mminfo = (MinMaxAggInfo *) lfirst(cell);
+
+				if (mminfo->aggfnoid == aggref->aggfnoid &&
+					equal(mminfo->target, curTarget->expr))
+					return (Node *) copyObject(mminfo->param);
+			}
+		}
+	}
+	return expression_tree_mutator(node, mutate_aggref_node, (void *) context);
+}
+
+/*
+ * replace_aggref_in_tlist
+ *
+ * If MinMaxAggPath is chosen, instead of running aggregate
+ * function we will execute subquery that we've generated. Since we
+ * use subquery we need to replace target list Aggref node with Param
+ * node. Param node passes output value from the subquery.
+ *
+ */
+void
+replace_aggref_in_tlist(MinMaxAggPath *minmaxagg_path)
+{
+	MutatorContext context;
+
+	context.mm_path = minmaxagg_path;
+
+	((Path *) minmaxagg_path)->pathtarget->exprs = (List *) mutate_aggref_node((Node *) ((Path *) minmaxagg_path)->pathtarget->exprs, (void *) &context);
+}
+
+/* Stores function id (FIRST/LAST) with proper comparison strategy */
+typedef struct FuncStrategy
+{
+	Oid			func_oid;
+	StrategyNumber strategy;
+} FuncStrategy;
+
+static Oid	first_last_arg_types[] = {ANYELEMENTOID, ANYOID};
+
+struct FuncStrategy first_func_strategy = {.func_oid = InvalidOid,.strategy = BTLessStrategyNumber};
+struct FuncStrategy last_func_strategy = {.func_oid = InvalidOid,.strategy = BTGreaterStrategyNumber};
+
+static Oid
+get_function_oid(char *name, int nargs, Oid arg_types[])
+{
+	FuncCandidateList func_candidates;
+
+	func_candidates = FuncnameGetCandidates(list_make1(makeString(name)), nargs, NIL, false, false, false);
+	while (func_candidates != NULL)
+	{
+		if (func_candidates->nargs == nargs && ts_function_types_equal(func_candidates->args, arg_types, nargs))
+			return func_candidates->oid;
+		func_candidates = func_candidates->next;
+	}
+	elog(ERROR, "failed to find function %s with %d args", name, nargs);
+}
+
+static FuncStrategy *
+initialize_func_strategy(FuncStrategy *func_strategy, char *name, int nargs, Oid arg_types[])
+{
+	func_strategy->func_oid = get_function_oid(name, nargs, arg_types);
+	return func_strategy;
+}
+
+static FuncStrategy *
+get_func_strategy(Oid func_oid)
+{
+	if (first_func_strategy.func_oid == InvalidOid)
+		initialize_func_strategy(&first_func_strategy, "first", 2, first_last_arg_types);
+	if (last_func_strategy.func_oid == InvalidOid)
+		initialize_func_strategy(&last_func_strategy, "last", 2, first_last_arg_types);
+	if (first_func_strategy.func_oid == func_oid)
+		return &first_func_strategy;
+	if (last_func_strategy.func_oid == func_oid)
+		return &last_func_strategy;
+	return NULL;
+}
+
+static bool
+contains_aggref_node(List *sortClause, List *targetList)
```

Comment:
This will only detect toplevel aggref nodes, to properly detect aggref nodes you need to use expression_tree_walker. Not sure why we are diverging from postgres minmax code with this check.
I'd like to see a couple tests for non-toplevel aggregations.

Reply:
This runs on any subplan and not only top level nodes. It is used to detect Aggref nodes used in order by (no nesting there so no need to recursively walk through the tree).The reason for doing this is explained in a code comment where method is being used.

Reply:
I though you've talked about nested queries.
Done.